File: src/simulatorV2/pySim/ED.py
import json
from json import JSONEncoder
from os import path

class ED():
    def __init__(self, num_docs, patient_rate, department_size, waiting_size, chance=None):
        self.erack = queue.PriorityQueue()
        self.time = 0
        self.DoctorList = []
        self.DispoList = [] # this will need to get split to admit/dc
        self.patient_rate = patient_rate ## patient rate in terms of patients per hour
        self.department_size = department_size
        self.waiting_size = waiting_size
        self.WR = queue.PriorityQueue(waiting_size)
        self.num_docs = num_docs
        for i in range(num_docs):
            self.DoctorList.append(Doctor(self, 0.75, 10, 8))

        self.Laboratory = Laboratory(self, 12)
        self.LWBSCount = 0
        self.chance = chance

        ## returns ([time], [waiting room vol], [department vol], [erack (to be seen)] )
        self.times = []
        self.waitRoomVolume = []
        self.departmentVolume = []
        self.erackVolume = []


    def get_time(self):
        """getter for children"""
        return self.time

    def dispoAdd(self, pt):
        """setter for dispolist - add pt"""
        self.DispoList.append(pt)

    def dispoPop(self, pt):
        """setter for dispolist - rem pt"""
        # version 2 will need to delete
        self.DispoList.remove(pt)

    def get_volume(self):
        total = self.erack.qsize()
        for doc in self.DoctorList:
            total += doc.get_patient_totals()
        return total

    def get_total_volume(self):
        return str(self.get_volume()) + " / " + str(self.department_size)

    def get_total_WR(self):
        """returns a string of the waiting room volume vs. capacity"""
        return str(self.WR.qsize()) + " / " + str(self.waiting_size)

    def generate_patient_prob(self):
        """generate a patient based on a poisson distribution, place it if there's room..."""
        #Probz
        probability = self.chance or np.random.uniform(0,1)
        if probability <= (self.patient_rate / 60):
            # a patient was generated
            if self.get_volume() < self.department_size:
                self.erack.put(Patient(self, 3))
            elif self.WR.qsize() < self.waiting_size:
                self.WR.put(Patient(self, 3))
            else:
                newpt = Patient(self, 3)
                #print("Patient", newpt.get_ID(), "left without being seen!")
                self.LWBSCount += 1

    def update_WR_erack(self):
        """check if patients can be brought from the WR to the erack"""
        if self.get_volume() < self.department_size:
            if not self.WR.empty():
                newpt = self.WR.get()
                self.erack.put(newpt)

    def send_patient_labs(self, patient):
        self.Laboratory.get_next_patient(patient)
        
    def update(self):
        #print("Time:", self.time, "  In waiting room:", self.get_total_WR(), "  In department:", self.get_total_volume(), "  To be seen:", self.erack.qsize())

        self.collectMetrics()

        self.generate_patient_prob()
        self.Laboratory.update()
        for doc in self.DoctorList:
            doc.update()
        for pt in self.DispoList:
            pt.update()
        self.update_WR_erack()

        self.time += 1
        #print()


    ## BELOW ADDED BY TJS
    def collectMetrics(self):
        self.times.append(self.time)
        self.waitRoomVolume.append(self.WR.qsize())
        self.departmentVolume.append(self.get_volume())
        self.erackVolume.append(self.erack.qsize())


    def run(self, duration=200):
        for i in range(duration):
            self.update()
            #if i == (duration - 1):
                #self.WriteJSON()
                ##print("Times: ", self.times)
                ##print("WR: ", self.waitRoomVolume)
                ##print("Dept: ", self.departmentVolume)
                #s#print("ERack: ", self.erackVolume)

    def WriteJSON(self):
        data = {}
        data['args'] = {
            'numDocs': self.num_docs,
            'patientRate': self.patient_rate,
            'departmentSize': self.department_size,
            'waitingSize': self.waiting_size,
            'chance': self.chance
        }

        data['ans'] = {
            'time': self.time,
            'times': self.times,
            'waitRoomVolume': self.waitRoomVolume,
            'departmentVolume': self.departmentVolume,
            'erackVolume': self.erackVolume,
        }

        with open(path.join(path.dirname(__file__), 'data.json'), 'r+') as outfile:
            jsonResult = json.load(outfile)
            temp = jsonResult['results']
            #print("temp: " + str(len(temp)))
            temp.append(data)
            #print("temp: " + str(len(temp)))
            outfile.seek(0)
            json.dump({"results": temp}, outfile, indent=2)
            outfile.truncate()





import scipy as sp
import numpy as np
import queue

from Doctor import Doctor
from Patient import Patient
from Laboratory import Laboratory

File: src/simulatorV2/tsSim/ED.ts
import { Doctor } from "./Doctor" 
import { Patient } from "./Patient"
import { Laboratory } from "./Laboratory" 
import { randomUniform } from "d3-random";
import { PriorityQueue } from "./queues/PriorityQueue";

export interface EDProps {
  numDocs: number;
  patientRate: number;
  departmentSize: number;
  waitingSize: number;
  chance?: number;
}

export function BernoulliTrial(probability: number, chance?: number) {
  if (chance === undefined) {
    chance = randomUniform(0, 1)()
  }

  return (chance <= probability);
}

export class ED {

  erack: PriorityQueue;
  time: number;
  DoctorList: Array<Doctor>;
  DispoList: Array<Patient>;
  patient_rate: number;
  department_size: number;
  waiting_size: number;
  WR: PriorityQueue;
  num_docs: number;
  Laboratory: Laboratory;
  LWBSCount: number;
  chance?: number

  // RETURNS
  times: Array<number>;
  waitRoomVolume: Array<number>;
  departmentVolume: Array<number>;
  erackVolume: Array<number>;


  constructor(numDocs: number, patientRate: number, departmentSize: number, waitingSize: number, chance?: number) {
    //set ids to 0
    Laboratory.ID = 0;
    Patient.ID = 0;
    Doctor.ID = 0;
    this.erack = new PriorityQueue(-1)
    this.time = 0;
    this.DoctorList = [];
    this.DispoList = []
    this.patient_rate = patientRate;
    this.department_size = departmentSize;
    this.waiting_size = waitingSize;
    this.WR = new PriorityQueue(-1)
    this.num_docs = numDocs;

    for (var i = 0; i < this.num_docs; i++) {
      this.DoctorList.push(new Doctor(this, 0.75, 10, 8))
    }

    this.Laboratory = new Laboratory(this, 12)
    this.LWBSCount = 0;
    this.chance = chance

    this.times = [];
    this.waitRoomVolume = [];
    this.departmentVolume = [];
    this.erackVolume =[];

  }

  get_time() {
    return this.time;
  }

  dispoAdd(pt: Patient) {
    this.DispoList.push(pt);
  }

  dispoPop(pt: Patient) {
    this.DispoList = this.DispoList.filter((
      val) => {
        return val.ID !== pt.ID
      }
    );
  }

  get_volume() {
    var total = this.erack.length;
    for (var i = 0; i < this.DoctorList.length; i++) {
        total += this.DoctorList[i].get_patient_totals()
    }
    return total
  }

  get_total_volume() {
    return this.get_volume() + " / " + this.department_size;
  }
  
  get_total_WR() {
    return this.WR.length + " / " + this.waiting_size;
  }

  generate_patient_prob() {
    //generate a patient based on a poisson distribution, place it if there's room..."""
    var probability = this.chance || randomUniform(0,1);
    if (probability <= (this.patient_rate / 60)) {
      // a patient was generated
      if (this.get_volume() < this.department_size) {
        this.erack.put(new Patient(this, 3));
      } else if (this.WR.length < this.waiting_size) {
        this.WR.put(new Patient(this, 3));
      } else {
        const newpt = new Patient(this, 3)
        //console.log("Patient", newpt.get_ID(), "left without being seen!")
        this.LWBSCount += 1
      }
    }
  }

  update_WR_erack() {
    /// check if patients can be brought from WR to the erack
    if (this.get_volume() < this.department_size) {
      if (!this.WR.empty()) {
        const newpt = this.WR.get();
        //console.log("erack update from WR", this.erack.length, "WR: ", this.WR.length)
        this.erack.put(newpt);
      }
    }
  }

  send_patient_labs(patient: Patient) {
    this.Laboratory.get_next_patient(patient);
  }


  update() {
    //console.log("Time: ", this.time, "  In waiting room: ", this.get_total_WR(), " In department: ", this.get_total_volume(), "to be seen: ", this.erack.length);
    this.collectMetrics()

    this.generate_patient_prob()
    this.Laboratory.update()
    for (var i = 0; i < this.DoctorList.length; i++) {
      this.DoctorList[i].update()
    }
    for (var i2 = 0; i2 < this.DispoList.length; i2++) {
      this.DispoList[i2].update()
    }
    this.update_WR_erack();
    this.time += 1;
    //console.log("\n");
  }

//addded by tjs
collectMetrics() {
  this.times.push(this.time);
  this.waitRoomVolume.push(this.WR.length);
  this.departmentVolume.push(this.get_volume());
  this.erackVolume.push(this.erack.length);
}


  run(duration: number = 200) {
    for (var i = 0; i < duration; i++) {
      this.update()
    }
    //console.log("\ntimes: ", this.times, "\nwr: ", this.waitRoomVolume, "\ndept: ", this.departmentVolume, "\nerack: ", this.erackVolume)
    return this;
  }
}


export function run(opts: EDProps) {
  return new ED(opts.numDocs, opts.patientRate, opts.departmentSize, opts.waitingSize, opts.chance).run();
}

export default run;
